### ISSUE TITLE
Visual bugs (0.14.1)

### ISSUE THREAD

**Issue by bsutherland** [bsutherland/JuceOPLVSTi#52]
@gdreyband Moving this here so as not to spam users in the OSX ticket.

I'm going overseas for 5 days but definitely interested in your ideas! Yes, please feel free to submit pull requests via github, code contributions would be much appreciated.

Converting all the SBI files to presets might be one way to solve that clunky UI. Nice idea.

-------+--;

Visual bugs I noticed (not sure if it's on the PC version):
-Selecting a checkbox creates a really noticeable white border around the entire option. (fixed in 0.14.2)
-Clicking on a DDL makes the green box around it disappear until an unrelated area is clicked.
-Load SBI instrument still opens window behind all other active DAW windows.
-When loading SBI files, the channel window doesn't reflect in-use channels (likely a deeper issue) (fixed in 0.14.1)

Another question, rather than messing with SBI files would you be interested in me creating a number of presets? A lot of VST/AUs provide a whole host of presets built into the plugin. Right now there's 14, but I'd be happy to create a lot more.

Also, if it's not too much hassle for you, I'm going to continue working on the GUI in my spare time to make it better. Right now there are some things bothering me.
-Attentuation slider (might be better placed to the right of the envelope), I see some chunky "step" motion when dragging it, which leads me to believe it could be shorter.
-Desire for custom elements for sliders and knobs to replace most DDLs.

I'm not going to foist any major changes on you, but I'm going to work on a couple of more drastic variations and see how far I can get before I present them to you. Once I figure out how to make good looking functional custom elements in JUCE, I'll provide some of that myself.

**Comment by gdreyband** [comment]
-When loading SBI files, the channel window doesn't reflect in-use channels (likely a deeper issue).
This seems to be working, it could have just been a JUCE bug.

**Comment by gdreyband** [comment]
Tested on 13.2 and 14.1 Mac (just to be sure)

**-When switching patches, toggleable values ([V]ibrato, [T]remolo, [S]ustain, [K]eyscale env. rate) don't reflect accurately in the GUI: there is no event issued to toggle off elements in the UI.**

Example: on Subdue Bass, no toggles are selected. Switching to next patch Tromba enables T|S|K1, and S2. Switch back, to Subdue Bass, all 4 values remain "on" visually but sound is still correct (same as original state).

**Comment by gdreyband** [comment]
On an unrelated note, I made a few modifications to the UI for my own use but I'm not going to make a code commit because they "break" break compatibility. 

They are: 
Reversed the slider values for ASDR
Made the Keyscale Attenuation combo box list in magnitude order (0, -1.5, -3, -6).

**Comment by bsutherland** [comment]
Hi,

Just got back today. For now:

Thanks for the bug reports.

- About the white outline, need some time to figure out how to fix that one. On the Juce forum Jules suggested using a debugger to find where that outline is getting painted. Haven't used a debugger with a VST before, but some info here: https://forum.juce.com/t/how-to-debug-the-audio-plugin-project/6423
- Didn't notice the issue with the dropdown outline, thanks for pointing that out
- About the load SBI button, I can't reproduce this at the moment (Windows / Renoise).. I wonder if this may be OS / DAW specific?
- About the GUI not reflecting parameters when switching patches (I guess patch == preset?), I think that might be the issue reported here too: https://github.com/bsutherland/JuceOPLVSTi/issues/28 .. if we go ahead with converting the SBI files to presets, that will definitely need fixing

About GUI changes / improvements

- I'd love to incorporate your custom slider style - definitely suits the look of the plugin better. Just need to figure out how to draw custom sliders in Juce. I'm guessing it's a matter of overriding methods in the LookAndFeel class, like I did for the checkboxes.
- Attenuation slider: not really sure what you mean about the step motion, but I'm open to layout changes. I'm not a fan of knobs in general though - I find them hard to use for fine adjustments, so I went with sliders intentionally for that reason. They do look good, and I know that they're pretty common in VSTs, but for me function takes precedence over form.

About SBIs -> presets

I can probably do this pretty easily. Would appreciate your thoughts on a couple of points. There are literally thousands of SBI files at the moment. I don't have that much experience with DAWs and VSTs to be honest, and I wonder how manageable this will be in most DAWs. Are presets searchable in most DAWs? Is there any way to arrange them into folders?

**Comment by gdreyband** [comment]
About the white outline, need some time to figure out how to fix that one. On the Juce forum Jules suggested using a debugger to find where that outline is getting painted. Haven't used a debugger with a VST before, but some info here: https://forum.juce.com/t/how-to-debug-the-audio-plugin-project/6423

Didn't notice the issue with the dropdown outline, thanks for pointing that out

**I think these two are related issues, basically it's visibly showing an alternate "in focus" state, when really the change in state should be obvious by either the toggle changing or the presense of the DDL. I'm not sure if it's a JUCE issue or specific to this VST**

About the load SBI button, I can't reproduce this at the moment (Windows / Renoise).. I wonder if this may be OS / DAW specific?

**Could be, in Logic on Mac, it always opens behind everything else but since it's a specialized window, it also gets annoying about me doing other things until the window is closed. Don't know if there's basically a "always bring to top" function that could be used.**

About the GUI not reflecting parameters when switching patches (I guess patch == preset?), I think that might be the issue reported here too: #28 .. if we go ahead with converting the SBI files to presets, that will definitely need fixing

**The only parameters that aren't reflected are the toggles, everything else seems to be correct, even things like keyscale attenuation, velocity sensitivity (I think, as none of the presets seem to use it but I'll test it myself to make sure).**

About GUI changes / improvements

I'd love to incorporate your custom slider style - definitely suits the look of the plugin better. Just need to figure out how to draw custom sliders in Juce. I'm guessing it's a matter of overriding methods in the LookAndFeel class, like I did for the checkboxes.

**Yea I'm trying to figure this out myself. I've just started digging in the code and am able to make some pretty superficial changes (like rearrange order of keyscale attenuation) but I haven't yet toyed much with custom graphics.**

Attenuation slider: not really sure what you mean about the step motion, but I'm open to layout changes.
**The attenuation slider I'm realizing is way too long, it would be better served by as a hybrid between what I have now and my alternate GUI, so a vertical slider to the right of ADSR instead of a hugely long horizontal slider below**

I'm not a fan of knobs in general though - I find them hard to use for fine adjustments, so I went with sliders intentionally for that reason. They do look good, and I know that they're pretty common in VSTs, but for me function takes precedence over form.

**I actually agree with you, but in this case I think the strongest case for linear sliders is really only in ADSR and attenuation (volume). A lot of other things I think would be served better by a different control: a two state toggle (pill or square shaped) for vib/trem depth since they only have two positions, a knob for feedback to differentiate it from the ADSR (since it only has 8 positions), and really anything but a DDL for base frequency / harmonic, but the harmonic changes are pivotal in sound design so I'm making a very strong plea for turning the frequency multiplier DDL into a slider at the very least. The other DDLs are not messed with quite as frequently so it's fine if they stay as they are for the time being.**

About SBIs -> presets

I can probably do this pretty easily. Would appreciate your thoughts on a couple of points. There are literally thousands of SBI files at the moment. I don't have that much experience with DAWs and VSTs to be honest, and I wonder how manageable this will be in most DAWs. Are presets searchable in most DAWs? Is there any way to arrange them into folders?

**I think what I've seen that would work best without requiring a lot of custom code (such as an in-synth patch browser, which many do have) is what occurs in VSTs like OBXd. Instead of factory presets being individual patches, they lead to named folders (bass, lead, pad, etc), each of which has contains a number of patches. I would advise against making every SBI into a patch. The better course of action would be to have a number of carefully chosen patches that show off what OPLVSTi can do, and make no mistake, it's actually a fairly powerful (if low-resolution) 2-OP FM synth. There's definitely some mileage to be had, and I actually meant more like I would create custom patches that I think sound good and send you a bunch of preset values the way that they are stored in the DAW.**

**Comment by gdreyband** [comment]
An example from Logic's built in FM synth EFM1:
![efm1](https://cloud.githubusercontent.com/assets/9372336/22658208/26f8b4e0-ec4e-11e6-8501-e95c5e77cf35.png)

and JUCEOPLVSTi for comparison.
![juceoplvsti](https://cloud.githubusercontent.com/assets/9372336/22658229/44163bd8-ec4e-11e6-96ac-2322f04c782f.png)

**Comment by bsutherland** [comment]
> Yea I'm trying to figure this out myself. I've just started digging in the code and am able to make some pretty superficial changes (like rearrange order of keyscale attenuation) but I haven't yet toyed much with custom graphics.

Try opening OPLLookAndFeel.h/cpp and adding some of these methods:

https://www.juce.com/doc/classLookAndFeel__V3#a7f4548429a6b816775a111cda8d655a1
https://www.juce.com/doc/classLookAndFeel__V2#a42ab374ce5878d6e3277fbd69da24f14
https://www.juce.com/doc/classLookAndFeel__V2#ad2a0d9eb1e8a3cee3917a3f0623c6e56

If you can't figure it out, let me know and I'll have a go.

> I actually agree with you, but in this case I think the strongest case for linear sliders is really only in ADSR and attenuation (volume). A lot of other things I think would be served better by a different control: a two state toggle (pill or square shaped) for vib/trem depth since they only have two positions, 

This makes sense, and might look nicer than a checkbox.

> a knob for feedback to differentiate it from the ADSR (since it only has 8 positions),

That's a good point too. With only 8 settings, and the fact that it's not a lesser used setting, maybe a knob would be OK. As long as it's reasonably large :)

> and really anything but a DDL for base frequency / harmonic, since the harmonic changes are pivotal in sound design so I'm making a very strong plea for turning the frequency multiplier DDL into a slider at the very least.

That is a very good point. As you say the frequencies are fundamental to an FM synth, and a DDL does require multiple clicks. Slider would probably make sense.

> Instead of factory presets being individual patches, they lead to named folders (bass, lead, pad, etc), each of which has contains a number of patches.

So it is possible! Sounds good, I'll look into it.

> I would advise against making every SBI into a patch. The better course of action would be to have a number of carefully chosen patches that show off what OPLVSTi can do.

This sounds great! I'd really appreciate this. If you're willing to set them up, and save them as SBIs, I can convert them easily into presets.

**Comment by gdreyband** [comment]
I've attached the files (graphics and source) for a newer UI, but it's only marginally different. I'm not 100% happy with it but it's a lot easier to scan. I've reached the limits of what I can accomplish directly in JUCE without giving myself a headache.

This should look a lot like the one I mocked up in the Feb 5th post above.

![screen shot 2017-04-25 at 6 14 42 pm](https://cloud.githubusercontent.com/assets/9372336/25413949/25468528-29e3-11e7-82b1-04118bf363b3.png)


[JuceOPLVSTi Master.zip](https://github.com/bsutherland/JuceOPLVSTi/files/956948/JuceOPLVSTi.Master.zip)
